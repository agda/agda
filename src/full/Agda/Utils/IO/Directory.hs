module Agda.Utils.IO.Directory
  ( copyDirContent
  )
where

import Control.Monad
import Control.Monad.Writer

import System.Directory
import System.FilePath
import Data.ByteString as BS





-- | @copyDirContent src dest@ recursively copies directory @src@ onto @dest@.
--
--   First, a to-do list of copy actions is created.
--   Then, the to-do list is carried out.
--
--   This avoids copying files we have just created again, which can happen
--   if @src@ and @dest@ are not disjoint.
--   (See issue #2705.)
--
copyDirContent :: FilePath -> FilePath -> IO ()
copyDirContent src dest = mapM_ performAction =<< do
  (`appEndo` []) <$> execWriterT (copyDirContentDryRun src dest)

-- | Action to be carried out for copying a directory recursively.
--
data CopyDirAction
  = MkDir    FilePath
      -- ^ Create directory if missing.
  | CopyFile FilePath FilePath
      -- ^ Copy file if changed.

-- | Perform scheduled 'CopyDirAction'.
--
performAction :: CopyDirAction -> IO ()
performAction = \case
  MkDir d           -> createDirectoryIfMissing True d
  CopyFile src dest -> copyIfChanged src dest

-- | @copyDirContentDryRun src dest@ creates a to-do list
--   for recursively copying directory @src@ onto @dest@.
--
copyDirContentDryRun :: FilePath -> FilePath -> WriterT (Endo [CopyDirAction]) IO ()
copyDirContentDryRun src dest = do
  tell $ Endo (MkDir dest :)
  chlds <- lift $ getDirectoryContents src
  forM_ chlds $ \ x -> do
    isDir <- lift $ doesDirectoryExist (src </> x)
    case isDir of
      _ | x == "." || x == ".." -> return ()
      True  -> copyDirContentDryRun (src </> x) (dest </> x)
      False -> tell $ Endo (CopyFile (src </> x) (dest </> x) :)

-- | @copyIfChanged src dst@ makes sure that @dst@ exists
--   and has the same content as @dst@.
--
copyIfChanged :: FilePath -> FilePath -> IO ()
copyIfChanged src dst = do
  exist <- doesFileExist dst
  if not exist then copyFile src dst else do
    new <- BS.readFile src
    old <- BS.readFile dst
    unless (old == new) $ copyFile src dst
