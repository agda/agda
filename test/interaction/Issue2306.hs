#!/usr/bin/env runhaskell

{-# LANGUAGE RecordWildCards #-}

import System.Directory

import RunAgda

file       = "Issue2306.agda"
firstCode  = unlines
               [ "B : Set"
               , "B = ?"
               ]
secondCode = unlines
               [ "A : Set"
               , "A = ?"
               ] ++
             firstCode

main :: IO ()
main = runAgda ["--no-libraries"]
         $ \(AgdaCommands { .. }) -> do

  -- Discard the first prompt.
  echoUntilPrompt

  -- Create the file.
  writeUTF8File file firstCode

  -- Load the file, and wait for Agda to complete.
  loadAndEcho file

  -- Edit the file.
  writeUTF8File file secondCode

  -- Check what the type of the goal is.
  send $
    command "goal_type" file (Just "Interactive Indirect")
            (Just "Normalised 0 noRange \"\"")

  -- Echo any output received from Agda.
  echoUntilPrompt

  return ()
