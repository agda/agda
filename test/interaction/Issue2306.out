(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue2306 (Issue2306.agda)." t)
(agda2-status-action "")
(agda2-info-action "*All Goals*" "?0 : Set" nil)
((last . 1) . (agda2-goals-action '(0)))
(agda2-status-action "")
(agda2-info-action "*Current Goal*" "Set" nil)
